### PR TITLE
docker_swarm_service: Sort lists when checking for changes

### PR DESCRIPTION
**Pull request by mcassaniti**
##### SUMMARY
When two lists are checked for changes in this module, the lists are reported changed when the order of the items is different. This PR resolves this issue.
##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
docker_swarm_service

### CODE REVIEW

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 1]
```diff
@@ -1759,7 +1764,7 @@ def compare(self, os):
         force_update = False
         if self.endpoint_mode is not None and self.endpoint_mode != os.endpoint_mode:
             differences.add('endpoint_mode', parameter=self.endpoint_mode, active=os.endpoint_mode)
-        if self.env is not None and self.env != (os.env or []):
+        if has_list_changed(self.env, os.env):
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py:1831]: The current behavior is:
 * if the user did not specify `env` (i.e. `self.env is None`), don't compare.
 * if the user did specify `env`, compare user's `env` to the service's `env`.
 * if the service's `env` is empty, treat it as `[]`: this simply means that no env variables have been set for the service.

So this does not ignore empty lists, but treats them correctly. (What it doesn't do is consider order - you're totally correct on that one.)

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 2]
```diff
@@ -1244,16 +1244,21 @@ def has_dict_changed(new_dict, old_dict):
...
     """
     if new_list is None:
         return False
     old_list = old_list or []
     if len(new_list) != len(old_list):
         return True
-    for new_item, old_item in zip(new_list, old_list):
+
+    if sort_lists:
+        zip_data = zip(sort(new_list), sort(old_list))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py]:
```suggestion
        zip_data = zip(sorted(new_list), sorted(old_list))
```

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py]: Also note that simply sorting does not necessarily give the correct result for lists of dicts. So this change could actually recreate/update services where nothing changed.

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py:1831]: A better version would be:
```suggestion
        if has_list_changed(self.env, os.env):
        if self.env is not None and has_list_changed(self.env, os.env or []):
```

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py:1831]: (Same for the comparisons below.)

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py (hunk 2 above)
Comment (mcassaniti) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py]: Fixed using sorted over sort. You're right about the list of dictionaries issue, and I don't have a good solution here. This should be an improvement over what's currently being performed.

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py (hunk 1 above)
Comment (mcassaniti) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py:1831]: > The current behavior is:
> 
>     * if the user did not specify `env` (i.e. `self.env is None`), don't compare.
Already implemented (line 1251,1252)
> 
>     * if the user did specify `env`, compare user's `env` to the service's `env`.
Done
> 
>     * if the service's `env` is empty, treat it as `[]`: this simply means that no env variables have been set for the service.
Already implemented (line 1253)
> 
> 
> So this does not ignore empty lists, but treats them correctly. (What it doesn't do is consider order - you're totally correct on that one.)


I believe all the list comparisons seemed to want the following:

  * If the new list is empty, skip comparing
  * If the old list is None, treat it as the empty list
  * Compare the lists. Most should be comparisons between sorted lists.

Can you point out any list comparisons where the above assumptions break? If they do I can adjust them.

**Comment by mcassaniti** [comment]
So it seems that sorting a list of dictionaries is not supported in Python3. That presents a problem, especially for something like the example below:

  * `new_list = [{'a': 1}, {'b': 2}]`
  * `old_list = [{'b': 2}, {'a': 1}]`

Mount options are a prime case of the example above. One option would be to provide a dictionary key to sort on if the key exists. Thoughts?

**Comment by mcassaniti** [comment]
I've added dictionary key sorting for lists of dictionaries. I've also got an integration environment operational and tested properly.

Can someone clarify why the list of networks is an ordered list? The test suite was explicitly testing that a re-order of the list of networks resulted in a change. I can't see a reason for this.

**Comment by felixfontein** [comment]
> Can someone clarify why the list of networks is an ordered list? The test suite was explicitly testing that a re-order of the list of networks resulted in a change. I can't see a reason for this.

No idea. @dariko @hannseman can you comment on this?

**Comment by hannseman** [comment]
@mcassaniti @felixfontein no that doesn't make sense. It's probably a case of bad copy and paste from another list-test.

**Comment by dariko** [comment]
@felixfontein @mcassaniti I have no idea too, so i tried investigating.

No help from the API docs (https://docs.docker.com/engine/api/v1.40/#operation/ServiceCreate), here network is just described as a list of objects :(.

To poke at the API, I tried running a sample playbook
`````
-   hosts: localhost
    tasks:
    -   docker_network:
            driver: overlay
            name: n1
    -   docker_network:
            driver: overlay
            name: n2
    -   docker_swarm_service:
            name: s1
            networks:
            -   n1
            -   n2
            image: debian
            args:
            -   sleep
            -   3600
    -   docker_swarm_service:
            name: s2
            networks:
            -   n2
            -   n1
            image: debian
            args:
            -   sleep
            -   3600
`````

and a sample script
`````
docker service create --name s3 --network n1 --network n2 --detach debian sleep 3600
docker service create --name s4 --network n2 --network n1 --detach debian sleep 3600
`````

And it looks like the container are created with the same interfaces order:
`````
user@local:~$ docker exec -it s1.1.w6mvkohm16o1i6e60jvs2cryw ip r
default via 172.19.0.1 dev eth2 
10.0.10.0/24 dev eth0 proto kernel scope link src 10.0.10.3 
10.0.11.0/24 dev eth1 proto kernel scope link src 10.0.11.3 
172.19.0.0/16 dev eth2 proto kernel scope link src 172.19.0.3 

user@local:~$ docker exec -it s2.1.5t2rd25pwlt20c2mk2yasbt9z ip r
default via 172.19.0.1 dev eth2 
10.0.10.0/24 dev eth0 proto kernel scope link src 10.0.10.6 
10.0.11.0/24 dev eth1 proto kernel scope link src 10.0.11.6 
172.19.0.0/16 dev eth2 proto kernel scope link src 172.19.0.4 

user@local:~$ docker exec -it s3.1.rcv2v3xeypa4q7pdosbe5jf58 ip r
default via 172.19.0.1 dev eth2 
10.0.10.0/24 dev eth0 proto kernel scope link src 10.0.10.8 
10.0.11.0/24 dev eth1 proto kernel scope link src 10.0.11.8 
172.19.0.0/16 dev eth2 proto kernel scope link src 172.19.0.5 

user@local:~$ docker exec -it s4.1.ooyz5pa6202r5tz7zpqmtjgz8 ip r
default via 172.19.0.1 dev eth2 
10.0.10.0/24 dev eth0 proto kernel scope link src 10.0.10.10 
10.0.11.0/24 dev eth1 proto kernel scope link src 10.0.11.10 
172.19.0.0/16 dev eth2 proto kernel scope link src 172.19.0.6 
`````

Checking the API responses I noticed that the services created by this module had their network informations in the `Spec.Networks` field, with the same order as they were created with
`````
user@local:~$ docker service inspect s1|jq '.[].Spec.Networks[].Target'
"vo7flqhh4sbkxc9fiqroadg93"
"qqmatahg6o1b5ojrf76uubzfi"
user@local:~$ docker service inspect s2|jq '.[].Spec.Networks[].Target'
"qqmatahg6o1b5ojrf76uubzfi"
"vo7flqhh4sbkxc9fiqroadg93"
`````

while the ones created by the CLI commands had the network informations in `Spec.TaskTemplate.Networks` and did not retain the network order specified:
`````
user@local:~$ sudo docker service inspect s3|jq '.[].Spec.TaskTemplate.Networks[].Target'
"qqmatahg6o1b5ojrf76uubzfi"
"vo7flqhh4sbkxc9fiqroadg93"
user@local:~$ sudo docker service inspect s4|jq '.[].Spec.TaskTemplate.Networks[].Target'
"qqmatahg6o1b5ojrf76uubzfi"
"vo7flqhh4sbkxc9fiqroadg93"
`````

**TLDR**
Even if all the container created either via CLI or via module with any networks order resulted in the same interface order in the container, I think that there might be some corner case (present or future) in which this information will become relevant.
Maybe we could:
- preserve the order when creating a new service
- ignore the order when updating a service 
this would allow us to manage the 'maybe' cases in the future with a `preserve_network_order` flag.

@hannseman 
That said, the fact that the module-created services have the `Network` field only directly under `Spec`, while the CLI-created ones only have it in `Spec.TaskTemplate` have me a tad perplexed.

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py:1831]: If the new list is `[]`, you should compare. The user explicitly requests an empty list in that case. If the new list is `None`, don't compare, the user is saying "I don't care".

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py:1831]: (But that's already what you implemented. So it should be fine.)

**Comment by felixfontein** [comment]
For standalone docker containers, network order on creation makes a difference (at least when creating via CLI or with `networks_cli_compatible: yes`): the first network decides the main network stack the container is connected to; options like MAC address are used for it.

I don't think that is relevant for swarms though, so I guess the order is not really relevant.

**Comment by hannseman** [comment]
> @hannseman
> That said, the fact that the module-created services have the Network field only directly under Spec, while the CLI-created ones only have it in Spec.TaskTemplate have me a tad perplexed.
> 

In Ansible >=2.9 and Docker API >=1.25 this module will use `Spec.TaskTemplate.Networks` instead of `Spec.Networks`. See https://github.com/ansible/ansible/pull/58961. Also see https://github.com/moby/moby/pull/39495#issuecomment-549979721. 

Let's open up a new issue if you think that this behaviour is wrong so that we don't derail this PR.

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 3]
```diff
@@ -1244,19 +1244,58 @@ def has_dict_changed(new_dict, old_dict):
...
+        The list may contain dictionaries, so use the sort key to handle them.
+        """
+
+        try:
+            return sorted(unsorted_list)
+        except TypeError:
+            if isinstance(unsorted_list[0], dict):
+                if not sort_key:
+                    raise Exception(
+                        "A sort key was not specified when sorting list"
```
Comment (hannseman) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py]: We should try and use single-quotes. There're some inconsistencies in this module but the majority of strings are single-quoted.

Comment (mcassaniti) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py]: Done.

File: test/units/modules/cloud/docker/test_docker_swarm_service.py [hunk 4]
```diff
@@ -169,60 +170,89 @@ def test_has_dict_changed(docker_swarm_service):
...
     assert not docker_swarm_service.has_list_changed(
-        [],
-        None
+        [{'a': 1}, {'a': 2}],
+        [{'a': 1}, {'a': 2}],
+        sort_key='a'
     )
+
+    if python_version_info[0] >= 3:
+        # This test does not work on Python 2
```
Comment (felixfontein) [inline test/units/modules/cloud/docker/test_docker_swarm_service.py]: Maybe it's better to modify `has_list_changed` so it gives the same results for Python 2 **and** Python 3?

**Review by felixfontein — COMMENTED** [review]
BTW, you also need a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs-how-to).

File: test/units/modules/cloud/docker/test_docker_swarm_service.py (hunk 4 above)
Comment (mcassaniti) [reply test/units/modules/cloud/docker/test_docker_swarm_service.py]: I've updated the code to make this consistent. The exception should only be thrown on a programming error.

**Comment by dariko** [comment]
@hannseman yes, sorry; I missed that PR

File: changelogs/fragments/63887-docker_swarm_service-sort-lists-when-checking-changes.yml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - docker_swarm_service: Sort lists when checking for changes
```
Comment (felixfontein) [inline changelogs/fragments/63887-docker_swarm_service-sort-lists-when-checking-changes.yml]:
```suggestion
  - "docker_swarm_service - Sort lists when checking for changes."
```

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 6]
```diff
@@ -1266,6 +1306,33 @@ def has_list_changed(new_list, old_list):
...
+    if len(new_networks) != len(old_networks):
+        return True
+
+    zip_data = zip(
+        sorted(new_networks, key=lambda k: k['id']),
+        sorted(old_networks, key=lambda k: k['id'])
+    )
+
+    for new_item, old_item in zip_data:
+        # Sort the aliases
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py:1326]: Please create shallow copies to avoid the original data to be modified:
```suggestion
        new_item = dict(new_item)
        old_item = dict(old_item)
        # Sort the aliases
```

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 7]
```diff
@@ -1266,6 +1306,33 @@ def has_list_changed(new_list, old_list):
...
+        if 'aliases' in new_item:
+            new_item['aliases'].sort()
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py]:
```suggestion
            new_item['aliases'] = sorted(new_item['aliases'])
```

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 8]
```diff
@@ -1266,6 +1306,33 @@ def has_list_changed(new_list, old_list):
...
+        if 'aliases' in old_item:
+            old_item['aliases'].sort()
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py]:
```suggestion
            old_item['aliases'] = sorted(old_item['aliases'])
```

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py (hunk 7 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_swarm_service.py]: Maybe even better:
```suggestion
            new_item['aliases'] = sorted(new_item['aliases'] or [])
```
To convert `None` to an empty list.

File: test/units/modules/cloud/docker/test_docker_swarm_service.py [hunk 9]
```diff
@@ -2,6 +2,7 @@
 __metaclass__ = type
 
 import pytest
+from sys import version_info as python_version_info
```
Comment (felixfontein) [inline test/units/modules/cloud/docker/test_docker_swarm_service.py]: Could you please remove this line?

**Review by felixfontein — COMMENTED** [review]
Looks good IMO, except some smaller things :)

File: test/units/modules/cloud/docker/test_docker_swarm_service.py (hunk 9 above)
Comment (mcassaniti) [reply test/units/modules/cloud/docker/test_docker_swarm_service.py]: Done.

File: lib/ansible/modules/cloud/docker/docker_swarm_service.py [hunk 10]
```diff
@@ -1266,6 +1306,35 @@ def has_list_changed(new_list, old_list):
...
+    )
+
+    for new_item, old_item in zip_data:
+        new_item = dict(new_item)
+        old_item = dict(old_item)
+        # Sort the aliases
+        if 'aliases' in new_item:
+            new_item['aliases'] = sorted(new_item['aliases'] or [])
+        if 'aliases' in old_item:
+            old_item['aliases'] = sorted(old_item['aliases'])
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_swarm_service.py]: Please also change it here:
```suggestion
            old_item['aliases'] = sorted(old_item['aliases'] or [])
```

**Comment by mcassaniti** [comment]
@felixfontein Thank you for the review. I apologise that there's now 20 commits for something that should be easy enough. I don't know what Ansible's processes are, but I'd suggest a squash commit.

Thanks again.

**Comment by felixfontein** [comment]
@mcassaniti thanks for fixing this! 20 commits is no problem, everything is squash-merged anyway :)

**Review by felixfontein — APPROVED** [review]
I think this is fine. @hannseman @dariko could you please take a look, in particular whether the correct values for `sort_lists` and `sort_keys` are used in `compare()`?

**Review by hannseman — APPROVED** [review]
Really good work! 👍

**Comment by felixfontein** [comment]
@mcassaniti thanks a lot for fixing this!
@hannseman thanks for reviewing this!

I'll create a backport PR to stable-2.9 so it will hopefully end up in 2.9.2. Do you think it should also be backported to stable-2.8?

**Comment by mcassaniti** [comment]
Thanks for all the feedback @felixfontein @hannseman
